src/coordinator/coordinator_server.cc

```diff
@@ -0,0 +1,441 @@
+      presumed_abort_timespec.tv_nsec);
+  return final_config;
+}
+
+absl::StatusOr<const std::string> GetTransactionId(
+    const std::string &client_transaction_id,
+    const std::string &client_address) {
+  std::array<unsigned char, 32> hash_value;
+  if (!HashSha256(absl::StrCat(client_transaction_id, "~", client_address),
+                  hash_value.data())) {
+    return absl::UnknownError(
+        "Could not hash client transaction id and address");
+  }
+  std::stringstream ss;
+  for (const auto &hash_byte : hash_value) {
+    ss << absl::StrCat(absl::Hex(hash_byte, absl::kZeroPad2));
+  }
+  return ss.str();
+}
+
+SplitTransaction SplitClientTransaction(
+    const common::Transaction &original_transaction) {
+  absl::flat_hash_map<std::string, TransactionWithNamespace>
+      transactions_by_namespace;
+  bool has_writes = false;
+  bool has_reads = false;
+  for (const auto &operation : original_transaction.ops()) {
+    TransactionWithNamespace &transaction_with_namespace =
+        transactions_by_namespace[operation.namespace_().address()];
+    transaction_with_namespace.ns = operation.namespace_();
+    *transaction_with_namespace.transaction.add_ops() = operation;
+    if (operation.has_put()) {
+      has_writes = true;
+    } else {
+      has_reads = true;
+    }
+    switch (transaction_with_namespace.transaction_type) {
+      case UNSET_TRANSACTION:
+        transaction_with_namespace.transaction_type =
+            operation.has_put() ? WRITE_ONLY_TRANSACTION
+                                : READ_ONLY_TRANSACTION;
+        break;
+      case READ_ONLY_TRANSACTION:
+        transaction_with_namespace.transaction_type =
+            operation.has_put() ? READ_WRITE_TRANSACTION
+                                : READ_ONLY_TRANSACTION;
+        break;
+      case WRITE_ONLY_TRANSACTION:
+        transaction_with_namespace.transaction_type =
+            operation.has_put() ? WRITE_ONLY_TRANSACTION
+                                : READ_WRITE_TRANSACTION;
+        break;
+      case READ_WRITE_TRANSACTION:
+        break;
+    }
+  }
+  std::vector<TransactionWithNamespace> split_transactions;
+  split_transactions.reserve(transactions_by_namespace.size());
+  for (const auto &transaction_with_namespace : transactions_by_namespace) {
+    split_transactions.push_back(transaction_with_namespace.second);
+  }
+  std::sort(split_transactions.begin(), split_transactions.end(),
+            [](const TransactionWithNamespace &t1,
+               const TransactionWithNamespace &t2) {
+              return t1.ns.address() < t2.ns.address();
+            });
+  return SplitTransaction{
+      .transactions = split_transactions,
+      .transaction_type = has_reads ? (has_writes ? READ_WRITE_TRANSACTION
+                                                  : READ_ONLY_TRANSACTION)
+                                    : WRITE_ONLY_TRANSACTION};
+}
+}  // namespace
+
+grpc::Status CoordinatorServer::CommitAtomicTransaction(
```

Comment:
Maybe we split this function into smaller helper methods to help readability.

Reply:
I moved out the part that sends the cohort prepare requests, and I think it looks better now. Thanks!